File: web/static/js/pages/backtest.js
// 回测页交互

async function runBacktest() {
    const tsCode = document.getElementById("bt-ts-code").value.trim();
    const days = parseInt(document.getElementById("bt-days").value) || 240;
    const stopLoss = parseFloat(document.getElementById("bt-stop-loss").value) || 0.07;
    const takeProfit = parseFloat(document.getElementById("bt-take-profit").value) || 0.15;
    const container = document.getElementById("backtest-result");
    document.getElementById("tune-result").innerHTML = "";
    container.innerHTML = '<p style="color:var(--text-dim)">回测中...</p>';

    try {
        const data = await API.backtest({
            ts_code: tsCode, days, stop_loss_pct: stopLoss, take_profit_pct: takeProfit,
        });

        const winRateColor = data.win_rate >= 0.5 ? "var(--red)" : data.win_rate >= 0.3 ? "var(--accent)" : "var(--text-dim)";
        let html = '<div style="background:var(--card-bg);border:1px solid var(--border);border-radius:8px;padding:1.5rem">';
        html += `<h3 style="color:var(--accent);margin-bottom:0.75rem">${data.ts_code} 回测结果</h3>`;
        html += '<div style="display:grid;grid-template-columns:repeat(auto-fill,minmax(150px,1fr));gap:1rem;margin-bottom:1rem">';
        html += `<div><div style="color:var(--text-dim);font-size:0.8rem">总交易次数</div><div style="font-size:1.5rem">${data.total_trades}</div></div>`;
        html += `<div><div style="color:var(--text-dim);font-size:0.8rem">盈利次数</div><div style="font-size:1.5rem;color:var(--red)">${data.win_trades}</div></div>`;
        html += `<div><div style="color:var(--text-dim);font-size:0.8rem">亏损次数</div><div style="font-size:1.5rem;color:var(--green)">${data.loss_trades}</div></div>`;
        html += `<div><div style="color:var(--text-dim);font-size:0.8rem">胜率</div><div style="font-size:1.5rem;color:${winRateColor}">${(data.win_rate * 100).toFixed(1)}%</div></div>`;
        html += `<div><div style="color:var(--text-dim);font-size:0.8rem">盈亏比</div><div style="font-size:1.5rem">${data.profit_factor.toFixed(2)}</div></div>`;
        html += `<div><div style="color:var(--text-dim);font-size:0.8rem">最大回撤</div><div style="font-size:1.5rem;color:var(--green)">${(data.max_drawdown * 100).toFixed(1)}%</div></div>`;
        html += `<div><div style="color:var(--text-dim);font-size:0.8rem">总收益率</div><div style="font-size:1.5rem;color:${data.total_return >= 0 ? "var(--red)" : "var(--green)"}">${(data.total_return * 100).toFixed(2)}%</div></div>`;
        html += '</div>';

        if (data.trades.length > 0) {
            html += '<table class="data-table"><thead><tr><th>买入日期</th><th>买入价</th><th>卖出日期</th><th>卖出价</th><th>收益%</th><th>持仓天</th><th>退出原因</th></tr></thead><tbody>';
            data.trades.forEach((t) => {
                const pnlColor = t.pnl_pct >= 0 ? "var(--red)" : "var(--green)";
                html += `<tr>
                    <td>${t.entry_date}</td><td>${t.entry_price.toFixed(2)}</td>
                    <td>${t.exit_date || "-"}</td><td>${t.exit_price ? t.exit_price.toFixed(2) : "-"}</td>
                    <td style="color:${pnlColor}">${(t.pnl_pct * 100).toFixed(2)}%</td>
                    <td>${t.hold_days}</td><td style="font-size:0.8rem">${t.exit_reason}</td>
                </tr>`;
            });
            html += '</tbody></table>';
        }
        html += '</div>';
        container.innerHTML = html;
    } catch (e) {
        container.innerHTML = `<p style="color:var(--red)">回测失败: ${e.message}</p>`;
    }
}

async function runTune() {
    const tsCode = document.getElementById("bt-ts-code").value.trim();
    const days = parseInt(document.getElementById("bt-days").value) || 240;
    const container = document.getElementById("tune-result");
    document.getElementById("backtest-result").innerHTML = "";
    container.innerHTML = '<p style="color:var(--text-dim)">参数调优中...</p>';

    const paramGrid = {
        stop_loss_pct: [0.03, 0.05, 0.07, 0.10, 0.15],
        take_profit_pct: [0.08, 0.10, 0.15, 0.20, 0.30],
    };

    try {
        const data = await API.tuneBacktest({
            ts_code: tsCode, param_grid: paramGrid, days, score_metric: "win_rate",
        });

        let html = '<div style="background:var(--card-bg);border:1px solid var(--border);border-radius:8px;padding:1.5rem">';
        html += `<h3 style="color:var(--accent);margin-bottom:0.75rem">参数调优结果</h3>`;
        html += `<p style="margin-bottom:0.5rem">最优参数: 止损=${data.best_params.stop_loss_pct}, 止盈=${data.best_params.take_profit_pct}, 胜率=${(data.best_score * 100).toFixed(1)}%</p>`;
        html += '<table class="data-table"><thead><tr><th>止损</th><th>止盈</th><th>胜率</th><th>总收益</th><th>最大回撤</th><th>交易数</th></tr></thead><tbody>';
        data.all_results.forEach((r) => {
            html += `<tr>
                <td>${r.params.stop_loss_pct}</td>
                <td>${r.params.take_profit_pct}</td>
                <td style="color:${r.score >= 0.5 ? "var(--red)" : "var(--text-dim)"}">${(r.score * 100).toFixed(1)}%</td>
                <td>${(r.total_return * 100).toFixed(2)}%</td>
                <td>${(r.max_drawdown * 100).toFixed(1)}%</td>
                <td>${r.total_trades}</td>
            </tr>`;
        });
        html += '</tbody></table></div>';
        container.innerHTML = html;
    } catch (e) {
        container.innerHTML = `<p style="color:var(--red)">调优失败: ${e.message}</p>`;
    }
}

File: api/routes/backtest.py
"""回测 + 调优路由"""

from fastapi import APIRouter
from pydantic import BaseModel
from api.schemas import (
    BacktestResultResponse, TuneResultResponse,
    ScreenResultResponse, StockScoreItem, TradeItem,
)

router = APIRouter(prefix="", tags=["回测"])


class BacktestRequest(BaseModel):
    ts_code: str
    days: int = 240
    stop_loss_pct: float = 0.07
    take_profit_pct: float = 0.15


class TuneRequest(BaseModel):
    ts_code: str
    param_grid: dict
    days: int = 240
    score_metric: str = "win_rate"


class HistoricalScreenRequest(BaseModel):
    date_range: dict
    criteria: dict = {}


@router.post("", response_model=BacktestResultResponse)
def run_backtest(req: BacktestRequest):
    """单次回测"""
    from core.backtest import backtest_strategy

    result = backtest_strategy(
        req.ts_code, days=req.days,
        stop_loss_pct=req.stop_loss_pct,
        take_profit_pct=req.take_profit_pct,
    )
    return BacktestResultResponse(
        ts_code=result.ts_code,
        total_trades=result.total_trades,
        win_trades=result.win_trades,
        loss_trades=result.loss_trades,
        win_rate=result.win_rate,
        profit_factor=result.profit_factor,
        max_drawdown=result.max_drawdown,
        avg_return=result.avg_return,
        total_return=result.total_return,
        trades=[
            TradeItem(
                ts_code=t.ts_code, entry_date=t.entry_date,
                entry_price=t.entry_price,
                exit_date=t.exit_date, exit_price=t.exit_price,
                pnl=t.pnl, pnl_pct=t.pnl_pct,
                hold_days=t.hold_days, exit_reason=t.exit_reason,
            )
            for t in result.trades
        ],
    )


@router.post("/tune", response_model=TuneResultResponse)
def tune_backtest(req: TuneRequest):
    """参数调优（网格搜索）"""
    from core.backtest import tune_params

    result = tune_params(
        req.ts_code, req.param_grid,
        days=req.days, score_metric=req.score_metric,
    )
    return TuneResultResponse(
        best_params=result.best_params,
        best_score=result.best_score,
        all_results=result.all_results,
    )


@router.post("/screener", response_model=ScreenResultResponse)
def historical_screen(req: HistoricalScreenRequest):
    """历史选股筛选"""
    from core.backtest import screen_historical

    end_date = req.date_range.get("end", "")
    strategies = req.criteria.get("strategies", [])
    min_score = req.criteria.get("min_score", 0)

    result = screen_historical(
        date=end_date, strategies=strategies, min_score=min_score,
    )
    return ScreenResultResponse(
        date=result.date,
        total_scanned=result.total_scanned,
        results=[
            StockScoreItem(
                ts_code=s.ts_code, name=s.name, score=s.score,
                b1_score=s.b1_score, trend_score=s.trend_score,
                volume_score=s.volume_score, risk_score=s.risk_score,
                reasons=list(s.reasons), warnings=list(s.warnings),
                rating=s.rating,
            )
            for s in result.results
        ],
    )
